### ISSUE TITLE
Custom response does not include headers set in dependencies

### ISSUE THREAD

**Issue by sk-** [tiangolo/fastapi#5498]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import fastapi


def set_custom_header(response: fastapi.Response) -> None:
    response.headers["X-Custom-Header"] = "Testing Response Dependencies"


app = fastapi.FastAPI(dependencies=[fastapi.Depends(set_custom_header)])


@app.get("/response")
def response():
    return {"status": "ok"}


@app.get("/custom_response")
def custom_response():
    return fastapi.responses.JSONResponse({"status": "ok"})
```

### Description

When an endpoint returns an explicit `Response` headers set in dependencies are not set. However, if one returns the plain data and relies on the default response then the headers are included.

With the example code above, I one fetches `/response` then the header `x-custom-header` will be set, but not when using the endpoint `/custom_response`.

```
$ curl localhost:8000/response -v
> GET /response HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< date: Sat, 15 Oct 2022 19:09:55 GMT
< server: uvicorn
< content-length: 15
< x-custom-header: Testing Responde dependencies
<
{"status":"ok"}


$ curl localhost:8000/custom_response -v
> GET /custom_response HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< date: Sat, 15 Oct 2022 19:09:58 GMT
< server: uvicorn
< content-length: 15
<
{"status":"ok"}
```

The part where the headers are added to the default response is here 

https://github.com/tiangolo/fastapi/blob/90fc4299d1d24a54dda8bf6f01cf3779ce6cf467/fastapi/routing.py#L263-L264

and here is the code where a custom response is returned

https://github.com/tiangolo/fastapi/blob/90fc4299d1d24a54dda8bf6f01cf3779ce6cf467/fastapi/routing.py#L235-L238

### Operating System

macOS

### FastAPI Version

0.79.0

### Python Version

Python 3.10.8

**Comment by jgould22** [comment]
This feels like expected behavior. 

You have edited the response object in the dependency and then created a fresh new response in your controller without the edited headers.

Anything you have done to the response object previously is thrown away in favor of your new response object.

If you wanted to be sure all responses get a header and not worry about the controller you could do it with an ASGI middle-ware which would add the header after the controller is run.

**Comment by sk-** [comment]
The problem with this is that now there are two ways of doing things depending on the context. Note also that it's not always trivial to change a dependency to a middleware, as you may want to have that only for certain paths.

There  are many suggestions, throughout the issue tracker and Stack Overflow recommending to use custom responses like `ORJson`. However, if you just change your controllers returns from a model to a response, then code that was working before won't work anymore.

IMHO, the behavior of both controllers should be equivalent:

```py
@app.get("/default_response", response_class=fastapi.responses.ORJSONResponse)
def default_response(response: fastapi.Response):
    return {"status": "ok"}

@app.get("/explicit_response")
def explicit_response(response: fastapi.Response):
    return fastapi.responses.ORJSONResponse({"status": "ok"})
```

Additionally, detecting this behavior is tricky, as you may have a test for your dependency that's using model returns and someone in your codebase changed the response type and voila you have a really hard to debug bug. This happened to us.

**Comment by sm-Fifteen** [comment]
That does make it sound like returning response objects should be warned against and the route decorator be modified to allow passing the return data directly to the response class (in order to skip validation or return pre-encoded data, which are the main usual reasons why you would return response objects instead of leaving it up to FastAPI) so this sort of workaround is never needed. The return types of route functions are already [intentionally left open-ended](https://github.com/tiangolo/fastapi/blob/f67b19f0f73ebdca01775b8c7e531e51b9cecfae/docs/en/docs/tutorial/response-model.md?plain=1#L45-L46), but the current workaround for #1224 and #1359 of just returning pre-instanciated responses never felt right to me api-wise because of how (like you said) they split the way return values are handled (and make `response_class` feel redundent at times).

**Comment by tiangolo** [comment]
Yep, as @jgould22 says, returning responses directly is a scape hatch for you to override anything FastAPI would do and take over directly: https://fastapi.tiangolo.com/advanced/response-directly/#return-a-response

They are not intended as the main use case. And yes, this means there are two ways to do things, do them through the normal ways, returning data directly, or taking over the wheel and doing things yourself. It's not great to have two ways to do things, but there's no other way to let you take the wheel, so we have to leave with that.

Maybe that section in the docs could include another note apart from saying that any data validation won't be done, and also say that no additional headers or background tasks set in other places will be included. If you think that would make things clearer for you, please make a PR there to the docs with that extra note. 🙏
